### DISCUSSION TITLE
Consider adding autofocus support and aligning better with HTML specs

### DISCUSSION THREAD

**Discussion by Chudesnov** [focus-trap/focus-trap#1021] [Ideas]
Currently the first tabbable node (using the `tabbable` package) is focused upon activating the trap unless specified otherwise via the `initialFocus` option:
https://github.com/focus-trap/focus-trap/blob/7b28c407dabb22b8e9728c8c2d0c461bfe746b0e/index.js#L792-L796
https://github.com/focus-trap/focus-trap/blob/7b28c407dabb22b8e9728c8c2d0c461bfe746b0e/index.js#L269-L289

This isn't how native dialogs worked up until recently (in most implementations they looked for the first _focusable_ one) nor does this behavior match the updated spec (albeit much closer to it) for dialogs as well as popovers, see https://github.com/whatwg/html/commit/a9f103c9f7bd09ef712990194638c75db1f50e3c

In order for focus-trap to become compatible, it would need to add an additional step before the existing one that checks for `autofocus` being set on either a focusable descendant of the dialog or the dialog itself (the latter is also supported by the updated spec I linked to above).

I'm creating this thread to understand whether the maintainers would welcome such a change (maybe being aligned with native HTML widgets that manage their focus is not a priority, or you'd like to avoid any breaking changes) and discuss the appropriate way of implementing it and communicating it to developers that use this library.

**Comment by stefcameron** [comment]
Thanks for starting this discussion! I'd much rather avoid breaking changes whenever possible. That's definitely an every-present consideration.

Focus-trap's initial focus behavior has been as it is for a very long time, and it isn't just for dialogs, although that's probably a common use. So at first glance, this sounds like another option to opt-in to new behavior which _could_ eventually be flipped around to be opt-out with a future major version (i.e. permission to break stuff if it makes things generally better).

There's definitely something to be said about not interfering with an element that has `autofocus` inside a trap container when the trap gets created. That specific part feels like it could be separate from the other part which would be adding support specifically-related to the newer `<dialog>` element.

Dialog element-specific could support could look like simply checking if:

1. There's only one container node; and
2. This container node is a `<dialog>`; and
3. The container node is included in the trap; and
4. There are no descendant nodes that can receive focus.

Instead of throwing the "your focus trap must have at least one tabbable element in it" error at (4), set focus to the container node -- or rather just let it have focus. Thing is, this behavior of falling back focus to the dialog itself sounds like a browser behavior which is probably fairly new, unless `<dialog>` elements were _always_ inherently focusable from inception of the element and all browsers allow focusing it already.

Another thing to check is whether setting the `autofocus` attribute on a node automatically makes it focusable (and possibly even tabbable), so just `<div autofocus>content</div>` is enough to get a focusable (and possibly tabbable) node that otherwise wouldn't be, or do you also need a `tabindex` attribute on it? Tabbable may need some adjusting depending on the answers here.

Yet another consideration is if the `<dialog>` element is inherently focusable, then Tabbable needs an enhancement to its `candidateSelectors` list:

https://github.com/focus-trap/tabbable/blob/1ad3d704c228968d32c1e9168dfd1c8714ac53b8/src/index.js#L6-L18

Also note that Tabbable defaults its `includeContainer` option to `false`, and Focus-trap uses the default unless the consumer specifies `createFocusTrap(container, { tabbableOptions: { includeContainer: true } });`

So if the container isn't included, and the container is the `<dialog>`, what should happen...?

Maybe Focus-trap's default behavior should change to include the container IIF the container is a `<dialog>` and the `tabbableOptions.includeContainer` option was _not_ explicitly set?

That's probably still a breaking change, but _maybe_ just a minor. Assuming we'd consider it breaking, then at least there would be a relatively easy way to opt-out and hopefully it would actually end-up being something that seems intuitive to most consumers so wouldn't cause too much grief. Otherwise, we avoid the breaking change and you can opt-in by setting `...includeContainer: true` when creating your trap (assuming the Tabbable tweaks are made).
